vpr/src/route/router_lookahead_cost_map.cpp

```diff
@@ -0,0 +1,408 @@
+#include "router_lookahead_cost_map.h"
+
+#include "router_lookahead_map_utils.h"
+#include "globals.h"
+#include "echo_files.h"
+
+#ifdef VTR_ENABLE_CAPNPROTO
+#    include "capnp/serialize.h"
+#    include "extended_map_lookahead.capnp.h"
+#    include "ndmatrix_serdes.h"
+#    include "mmap_file.h"
+#    include "serdes_utils.h"
+#endif
+
+// Distance penalties filling are calculated based on available samples, but can be adjusted with this factor.
+static constexpr float PENALTY_FACTOR = 1.f;
+static constexpr float PENALTY_MIN = 1e-12f;
+
+// also known as the L1 norm
+static int manhattan_distance(const vtr::Point<int>& a, const vtr::Point<int>& b) {
+    return abs(b.x() - a.x()) + abs(b.y() - a.y());
+}
+
+template<class T>
+static constexpr const T& clamp(const T& v, const T& lo, const T& hi) {
+    return std::min(std::max(v, lo), hi);
+}
+
+template<typename T>
+static vtr::Point<T> closest_point_in_rect(const vtr::Rect<T>& r, const vtr::Point<T>& p) {
+    if (r.empty()) {
+        return vtr::Point<T>(0, 0);
+    } else {
+        return vtr::Point<T>(clamp<T>(p.x(), r.xmin(), r.xmax() - 1),
+                             clamp<T>(p.y(), r.ymin(), r.ymax() - 1));
+    }
+}
+
+// build the segment map
```

Comment:
This should allow a faster lookup to get the segment id. I am unsure of the run-time impact of accessing this information through a subroutine instead of having a pre-computed map. Run-time could be affected during the get_map_cost procedure which should be kept as low as possible.

Reply:
I was also curious about this, but didn't try removing it.